services/shhext/service.go

```diff
@@ -251,3 +276,323 @@ func (s *Service) Stop() error {
 	s.tracker.Stop()
 	return nil
 }
+
+// JoinPublicChats join a set of public chats deriving the key from the chat name
+func (s *Service) JoinPublicChats(chats []string) ([]string, error) {
+
+	var response []string
+	if s.loadedFilters == nil {
+		s.loadedFilters = make(map[string]LoadedFilter)
+	}
+
+	for _, chatID := range chats {
+		symKeyID, err := s.w.AddSymKeyFromPassword(chatID)
+		if err != nil {
+			return nil, err
+		}
+
+		symKey, err := s.w.GetSymKey(symKeyID)
+		if err != nil {
+			return nil, err
+		}
+
+		topic := chat.ToWhisperTopic(chatID)
+		topics := [][]byte{topic[:]}
+
+		filterOpts := &whisper.Filter{
+			KeySym:   symKey,
+			Topics:   topics,
+			AllowP2P: true,
+			Messages: make(map[common.Hash]*whisper.ReceivedMessage),
+		}
+		filter, err := s.w.Subscribe(filterOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		loadedFilter := LoadedFilter{
+			SymKeyID: symKeyID,
+			FilterID: filter,
+		}
+
+		s.loadedFilters[chatID] = loadedFilter
+		response = append(response, filter)
+	}
+	return response, nil
+}
+
+// LeavePublicChats leaves set of public chats
+func (s *Service) LeavePublicChats(chats []string) error {
+	for _, chatID := range chats {
+		loadedFilter, present := s.loadedFilters[chatID]
+		if !present {
+			continue
+		}
+
+		s.w.DeleteSymKey(loadedFilter.SymKeyID)
+
+		err := s.w.Unsubscribe(loadedFilter.FilterID)
+		if err != nil {
+			return err
+		}
+
+		delete(s.loadedFilters, chatID)
+
+	}
+	return nil
+}
+
+// Post shamelessly copied from whisper codebase with slight modifications.
+func (s *Service) Post(ctx context.Context, req whisper.NewMessage) (hash hexutil.Bytes, err error) {
+	hash, err = s.publicAPI.Post(ctx, req)
+	if err == nil {
+		var envHash common.Hash
+		copy(envHash[:], hash[:]) // slice can't be used as key
```

Comment:
Can we fix this fragment? This comment does not make sense and I would check if `common.Hash` can't be initialized directly from a byte slice instead of using `copy`.

Reply:
Not sure, 
basically it is used for the key of a map,
`common.Hash` is basically an `[32]byte`, https://github.com/ethereum/go-ethereum/blob/3e8184006189b57e2d998313ef0fa75429b0c5bc/common/types.go#L47 , I have changed it to use `setBytes` https://github.com/ethereum/go-ethereum/blob/3e8184006189b57e2d998313ef0fa75429b0c5bc/common/types.go#L109, is that ok?

Reply:
Yes, it's better. Looks good!